lib/mutgen.c

```diff
@@ -182,70 +198,104 @@ mutgen_init_allocator(mutgen_t *self, tsk_table_collection_t *tables)
 }
 
 static int MSP_WARN_UNUSED
-mutgen_add_mutation(mutgen_t *self, node_id_t node, double position,
-        const char *ancestral_state, const char *derived_state)
+mutgen_add_mutation(mutgen_t *self, node_id_t node, double position, double time,
+        const char *ancestral_state, const char *derived_state,
+        avl_node_t* avl_node)
 {
     int ret = 0;
-    tsk_site_t *site = tsk_blkalloc_get(&self->allocator, sizeof(*site));
+    bool new_site = (avl_node == NULL);
+    site_list_t *site_list;
+    tsk_site_t *site;
+    mutation_list_t **insert_after;
+    mutation_list_t *mutation_list = tsk_blkalloc_get(&self->allocator, sizeof(*mutation_list));
     tsk_mutation_t *mutation = tsk_blkalloc_get(&self->allocator, sizeof(*mutation));
-    avl_node_t* avl_node = tsk_blkalloc_get(&self->allocator, sizeof(*avl_node));
-
-    if (site == NULL || mutation == NULL || avl_node == NULL) {
+    if (mutation == NULL || mutation_list == NULL) {
         ret = MSP_ERR_NO_MEMORY;
         goto out;
     }
-    memset(site, 0, sizeof(*site));
+    if (new_site) {
+        avl_node = tsk_blkalloc_get(&self->allocator, sizeof(*avl_node));
+        site_list = tsk_blkalloc_get(&self->allocator, sizeof(*site_list));
+        site = tsk_blkalloc_get(&self->allocator, sizeof(*site));
+        if (site == NULL || site_list == NULL || avl_node == NULL) {
+            ret = MSP_ERR_NO_MEMORY;
+            goto out;
+        }
+        site_list->site = site;
+        site_list->mutations_head = NULL;
+        site_list->mutations_length = 0;
+
+        memset(site, 0, sizeof(*site));
+        site->position = position;
+        site->ancestral_state = ancestral_state;
+        site->ancestral_state_length = 1;
+
+        avl_init_node(avl_node, site_list);
+        avl_node = avl_insert_node(&self->sites, avl_node);
+        assert(avl_node != NULL);
+    } else {
+        site_list = (site_list_t *) avl_node->item;
+        site = site_list->site;
+        assert(site->position == position);
+    }
+
     memset(mutation, 0, sizeof(*mutation));
-    site->position = position;
-    site->ancestral_state = ancestral_state;
-    site->ancestral_state_length = 1;
     mutation->derived_state = derived_state;
     mutation->derived_state_length = 1;
     mutation->node = node;
     mutation->parent = TSK_NULL;
-    site->mutations = mutation;
-    site->mutations_length = 1;
 
-    avl_init_node(avl_node, site);
-    avl_node = avl_insert_node(&self->sites, avl_node);
-    assert(avl_node != NULL);
+    memset(mutation_list, 0, sizeof(*mutation_list));
+    mutation_list->mutation = mutation;
+    mutation_list->time = time;
+    insert_after = &(site_list->mutations_head);
```

Comment:
This seems more complicated than it needs to be... What about a function?

```c
static void 
insert_mutation(site_t *site, mutation_list_t *new)
{
    mutation_list_t *prev, *u;

   
           prev = NULL;
           u = site->head;
           while (u != NULL && u->time < new->time) {
                 prev = u;
                 u = u->next;
            } 
            if (prev != NULL) {
                prev->next = new;
            } 
            new->next = u;
    }
}
```

Or something like this?